### PR TITLE
Tweaks to update_graph (backport from #8185)

### PR DESCRIPTION
**Pull request by crusaderky**
Miscellaneous non-controversial tweaks to update_graph, backported from #8185 (which may not happen for some time).

Notably, if update_graph fails this PR notifies the client who invoked update_graph in addition to all other clients in who_wants.
I didn't write a unit test as I'm unsure of how to trigger this use case without #8185.

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -8724,26 +8742,28 @@ def _materialize_graph(
...
+        # referencing themselves.
+        if not iskey(v) or v != k:
+            dsk2[k] = v
+
     dsk = dsk2
 
     # - Add in deps for any tasks that depend on futures
     for k, futures in fut_deps.items():
-        dependencies[k].update(f.key for f in futures)
+        dependencies[k].update(f.key for f in futures if f.key != k)
```
Comment (crusaderky) [inline distributed/scheduler.py:8760]: This is from @fjetter and I'm unclear about the use case for it...

Rest of the changes to the function are non-functional.

**Review by hendrikmakait — COMMENTED** [review]
Thanks, @crusaderky! CI failure is unrelated. There's one mysterious change in here that I'd like for us to understand better (cc @fjetter), but that's non-blocking.

**Review by hendrikmakait — APPROVED** [review]

File: distributed/scheduler.py (hunk 1 above)
Comment (crusaderky) [reply distributed/scheduler.py:8760]: I'm going to revert it to unblock this PR. Once @fjetter responds, I'll potentially write a dedicated PR (with unit test to demonstrate the edge case).

Comment (crusaderky) [inline distributed/scheduler.py:8760]:
```suggestion
        dependencies[k].update(f.key for f in futures)
```
